BIDebugEngine/BIDebugEngine/Debugger.cpp

```diff
@@ -445,6 +393,41 @@ void Debugger::onScriptHalt(GameState* gs) {
     hAction.execute(this, nullptr, DebuggerInstructionInfo{ nullptr,(RV_VMContext*)gs->get_vm_context(), gs });
 }
 
+namespace {
+    struct LastInstructionNameVisitor
+    {
+        void visit(const CallStackItemSimple* stackItem, int) {
+            lastInstructionName = stackItem->_instructions.get(stackItem->_currentInstruction - 1)->get_name();
+        }
+
+        void visit(const CallStackItemData* stackItem, int) {
+            lastInstructionName = stackItem->_code->instructions.get(stackItem->_ip - 1)->get_name();
+        }
+
+        void visit(const CallStackItemArrayForEach* stackItem, int) {}
+        void visit(const CallStackItemArrayCount* stackItem, int) {}
+        void visit(const CallStackItemForBASIC* stackItem, int) {}
+        void visit(const CallStackItemApply* stackItem, int) {}
+        void visit(const CallStackItemConditionSelect* stackItem, int) {}
+        void visit(const CallStackItemArrayFindCond* stackItem, int) {}
+        void visit(const CallStackItemFor* stackItem, int) {}
+        void visit(const CallStackRepeat* stackItem, int) {}
+
+        r_string lastInstructionName;
+    };
+
+    bool allowStepInto(const DebuggerInstructionInfo& instructionInfo, int lastStepFrame)
```

Comment:
Why is this in a anon namespace, instead of being a Debugger member function?

Reply:
Why would it be a Debugger member function? Its an internally used method only used in the cpp (in fact only in the function below it). Only reason for it to be a member function is if it needs access to internal state of the Debugger, or is part of the public interface.

Reply:
> Only reason for it to be a member function is if it needs access to internal state of the Debugger, or is part of the public interface.

Or if it belongs to it.
A static private member still belongs to the debugger, but is not really part of the interface and doesn't need to be.

Reply:
I will make this change if you want, it's wrong though. Classes aren't modules just for organizing related code, and implementation details should be constrained in the most private way possible.